File: src/sdks/nodejs/src/domain/sso-session-expired-error.ts
/**
 * Thrown when the AWS SSO session for the active profile could not be loaded
 * or has expired, so credentials cannot be resolved.
 */
export class SsoSessionExpiredError extends Error {
  readonly profileName?: string;
  readonly cause?: unknown;

  constructor(profileName?: string, cause?: unknown) {
    super(SsoSessionExpiredError.buildMessage(profileName));
    this.name = 'SsoSessionExpiredError';
    this.profileName = profileName;
    this.cause = cause;
  }

  private static buildMessage(profileName?: string): string {
    if (profileName) {
      return (
        `Your AWS SSO session for profile '${profileName}' could not be ` +
        `loaded or has expired. Run: aws sso login --profile ${profileName} ` +
        'and then re-run this command.'
      );
    }
    return (
      'Your AWS SSO session could not be loaded or has expired. Run: ' +
      'aws sso login and then re-run this command.'
    );
  }
}

const SSO_SESSION_EXPIRED_ERROR_NAMES = new Set([
  'TokenProviderError',
  'TokenRefreshRequired',
]);

/**
 * Detects whether an unknown error represents an expired or unresolved AWS SSO
 * session, by inspecting its AWS error name.
 */
export function isSsoSessionExpiredError(error: unknown): boolean {
  if (typeof error !== 'object' || error === null || !('name' in error)) {
    return false;
  }
  const name = String((error as { name?: unknown }).name);
  return SSO_SESSION_EXPIRED_ERROR_NAMES.has(name);
}

File: src/sdks/python/envilder/domain/sso_session_expired_error.py
"""Detection and typed error for expired or unresolved AWS SSO sessions."""

from __future__ import annotations

_SSO_SESSION_EXPIRED_EXCEPTION_NAMES = {
    "TokenRetrievalError",
    "UnauthorizedSSOTokenError",
    "SSOTokenLoadError",
}


def _build_message(profile_name: str | None) -> str:
    if profile_name:
        return (
            f"Your AWS SSO session for profile '{profile_name}' could not "
            f"be loaded or has expired. Run: aws sso login --profile "
            f"{profile_name} and then re-run this command."
        )
    return (
        "Your AWS SSO session could not be loaded or has expired. Run: "
        "aws sso login and then re-run this command."
    )


class SsoSessionExpiredError(Exception):
    """Raised when the AWS SSO session could not be loaded or has expired.

    Triggered when boto3 cannot resolve an SSO token for the active
    profile (for example, the cached SSO session expired). The message
    explains how to refresh the session with ``aws sso login`` and retry.
    When a profile is configured, the login hint is scoped to it.

    Example::

        from envilder import Envilder, SsoSessionExpiredError

        try:
            Envilder.resolve_file("envilder.json")
        except SsoSessionExpiredError as error:
            print(error)  # explains how to run aws sso login and retry
    """

    def __init__(self, profile_name: str | None = None) -> None:
        super().__init__(_build_message(profile_name))
        self.profile_name = profile_name


def is_sso_session_expired_error(error: BaseException) -> bool:
    """Detect whether an exception indicates an expired AWS SSO session.

    Relies only on the stable exception class name, mirroring the AWS
    token-resolution taxonomy. Message text is intentionally not
    inspected because its wording is unstable across SDK versions.

    Args:
        error: The exception raised by a boto3/botocore call.

    Returns:
        ``True`` when the error indicates an expired or unresolved SSO
        session.
    """
    return type(error).__name__ in _SSO_SESSION_EXPIRED_EXCEPTION_NAMES
